### Question 1


What's the version of Pandas that you installed?



In [1]:
import pandas as pd
print(f'Pandas version is {pd.__version__}')

Pandas version is 2.0.3


### Question 2

For this homework, we'll use the California Housing Prices dataset. Download it from here.

You can do it with wget:
```console
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
```    
Or just open it with your browser and click "Save as...".

Now read it with Pandas. How many columns are in the dataset?


In [2]:
housing = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv')

In [4]:
print(f'There are {len(housing.columns)} columns')

There are 10 columns


### Question 3

Which columns in the dataset have missing values?


In [6]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [10]:
housing.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

total_bedrooms column has 207 missing values

### Question 4
How many unique values does the ocean_proximity column have?


In [15]:
print(f'There are {len(housing.ocean_proximity.unique())} unique values')

There are 5 unique values


### Question 5
What's the average value of the median_house_value for the houses located near the bay?

In [20]:
near_bay_label = housing.ocean_proximity.unique()[0]
av_value = housing[housing.ocean_proximity == near_bay_label]['median_house_value'].mean()

In [21]:
print(f'Average value is {av_value}')

Average value is 259212.31179039303


### Question 6

* Calculate the average of total_bedrooms column in the dataset.
* Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
* Now, calculate the average of total_bedrooms again.
* Has it changed?

In [22]:
av_total_bedrooms = housing.total_bedrooms.mean()

In [27]:
av_total_bedrooms_with_filled_na = housing.total_bedrooms.fillna(av_total_bedrooms).mean()

In [28]:
av_total_bedrooms, av_total_bedrooms_with_filled_na

(537.8705525375618, 537.8705525375617)

Miracle, no changes.

### Question 7


* Select all the options located on islands.
* Select only columns housing_median_age, total_rooms, total_bedrooms.
* Get the underlying NumPy array. Let's call it X.
* Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
* Compute the inverse of XTX.
* Create an array y with values [950, 1300, 800, 1000, 1300].
* Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
* What's the value of the last element of w?

In [36]:
island_label = housing.ocean_proximity.unique()[-1:][0]
housinng_island = housing[housing.ocean_proximity == island_label]
X = housinng_island[['housing_median_age','total_rooms', 'total_bedrooms']].to_numpy()

In [46]:
import numpy as np
XTX = np.matmul(X.T, X)

In [47]:
inverse_XTX = np.linalg.inv(XTX)

In [41]:
y = np.array([950, 1300, 800, 1000, 1300])

In [51]:
w = np.matmul(np.matmul(inverse_XTX, X.T), y)

In [55]:
print(f'The last element is {w[-1:]}')

The last element is [5.69922946]
